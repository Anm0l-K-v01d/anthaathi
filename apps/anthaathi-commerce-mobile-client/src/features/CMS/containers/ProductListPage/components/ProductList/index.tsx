import * as React from 'react';
import {Avatar, Text, Title} from 'react-native-paper';
import {Image, Pressable, View, VirtualizedList} from 'react-native';
import Entypo from 'react-native-vector-icons/Entypo';
import {useResponsiveValue} from '../../../../utils/useResponsiveValue';
import {useIntl} from 'react-intl';
import {ProductListPageComponentType} from '../../../../types/common';

export interface ProductProps {
  name: string;
  description?: string;
  price: number;
  currency: string;
  image: string;
  weight_unit: string;
  packaging: string;
  key: string;
  notes: string;
}

export interface ProductListProps {
  products: ProductProps[];
  handlePress: () => void;
}

export default function ProductList({products, handlePress}: ProductListProps) {
  const itemHeight = useResponsiveValue([160, 200, 220, 280]);
  const itemWidth = useResponsiveValue([170, 190, 210, 270]);
  const productSplitted: ProductProps[][] = React.useMemo(() => {
    return products.reduce<ProductProps[][]>(
      (previousValue, currentValue, currentIndex) => {
        if (currentIndex % 2) {
          const newArray: ProductProps[][] = [...previousValue];
          newArray[previousValue.length - 1].push(currentValue);
          return newArray;
        } else {
          return [...previousValue, [currentValue]];
        }
      },
      [],
    );
  }, [products]);

  return (
    <View style={{marginTop: 10}}>
      <VirtualizedList<ProductProps[]>
        data={productSplitted}
        contentContainerStyle={{paddingBottom: 100}}
        testID="productList"
        // initialNumToRender={4}
        // horizontal
        renderItem={({item}) => (
          <ItemRendererColumn
            item={item}
            itemHeight={itemHeight}
            itemWidth={itemWidth}
            handlePress={handlePress}
          />
        )}
        getItemCount={() => productSplitted.length}
        keyExtractor={(item, index) => item?.[0]?.key || index + ''}
        getItem={(res, index) => res[index]}
      />
    </View>
  );
}

function ItemRendererColumn({
  item,
  itemHeight,
  itemWidth,
  handlePress,
}: {
  item: ProductProps[];
  itemHeight: number;
  itemWidth: number;
  handlePress: () => void;
}) {
  return (
    <View
      style={{
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-between',
        marginBottom: 15,
        marginHorizontal: 10,
      }}>
      {item.map(element => (
        <ItemRenderer
          key={element.key}
          item={element}
          itemHeight={itemHeight}
          itemWidth={itemWidth}
          handlePress={handlePress}
        />
      ))}
    </View>
  );
}

function ItemRenderer({
  item,
  itemHeight,
  itemWidth,
  handlePress,
}: {
  item: ProductProps;
  itemHeight: number;
  itemWidth: number;
  handlePress: () => void;
}) {
  const intl = useIntl();
  return (
    <View
      style={{
        marginVertical: 5,
        width: '48%',
        borderColor: item.name === 'Dabbas Dates' ? '#008D3E' : '#e7e7e7',
        backgroundColor: '#f0f0f0',
        borderWidth: 1,
        borderRadius: 12,
      }}
      key={item.key}>
      <View>
        <Pressable
          onPress={handlePress}
          onLongPress={() => {
            console.log('long press product add');
          }}>
          <View style={{height: itemHeight, width: '100%'}}>
            <Image
              style={{
                borderTopLeftRadius: 12,
                borderTopRightRadius: 12,
                height: itemHeight,
                width: '99.9%',
                paddingHorizontal: '0.1%',
                backgroundColor: '#fff',
              }}
              source={{
                uri: item.image,
              }}
            />
            {item.name === 'Dabbas Dates' && (
              <>
                <Pressable
                  onPress={() => {
                    console.log('press product minus');
                  }}
                  style={{
                    position: 'absolute',
                    left: 5,
                    top: 5,
                  }}>
                  <Avatar.Icon
                    icon="minus"
                    size={32}
                    style={{
                      backgroundColor: 'red',
                    }}
                  />
                </Pressable>
                <Pressable
                  onPress={() => {
                    console.log('press add product');
                  }}
                  style={{
                    position: 'absolute',
                    right: 5,
                    top: 5,
                  }}>
                  <Avatar.Text label={'1'} size={32} />
                </Pressable>
              </>
            )}
          </View>
          <View
            style={{
              alignItems: 'flex-start',
              backgroundColor: '#f0f0f0',
              paddingVertical: 5,
              paddingHorizontal: 8,
              borderBottomLeftRadius: 12,
              borderBottomRightRadius: 12,
            }}>
            <View
              style={{
                flexDirection: 'row',
                alignItems: 'flex-start',
                justifyContent: 'space-between',
                width: itemWidth,
              }}>
              <Title style={{fontSize: 12, marginRight: 5, fontWeight: 'bold'}}>
                {item.name}
              </Title>
              <Entypo
                name="info-with-circle"
                color="#364A15"
                size={18}
                style={{marginTop: 5}}
              />
            </View>
            <Text style={{color: '#808080', fontSize: 12, fontWeight: '400'}}>
              Dorne
            </Text>
            <View
              style={{
                flexDirection: 'row',
                marginTop: 5,
                alignItems: 'center',
                justifyContent: 'center',
              }}>
              <Text style={{color: '#008D3E', fontSize: 14, fontWeight: '400'}}>
                {intl.formatNumber(item.price, {
                  style: 'currency',
                  currency: item.currency,
                })}
              </Text>
              <Text style={{color: '#808080', fontSize: 12, fontWeight: '400'}}>
                {' / ' + item.packaging}
              </Text>
            </View>

            {/* <Text style={{color: '#808080', fontSize: 12, fontWeight: '400'}}>
              {item.notes}
            </Text> */}
          </View>
        </Pressable>
      </View>
    </View>
  );
}

export const ProductListCMSInput = {
  _component: ProductListPageComponentType.ProductList,
  component: ProductList,
};
