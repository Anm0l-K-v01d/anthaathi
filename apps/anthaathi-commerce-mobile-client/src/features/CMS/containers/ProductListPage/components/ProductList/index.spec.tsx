import React from 'react';
import 'react-native';
import {render} from '@testing-library/react-native';
import ProductList from './index';
import {IntlProvider} from 'react-intl';

import locale from '../../../../../../compiled-locales/en-US.json';
import {ThemeProvider} from 'react-native-paper';
import {RecoilRoot} from 'recoil';

describe('ProductList', () => {
  it('should render the item', function () {
    const temp = render(
      <ThemeProvider>
        <RecoilRoot>
          <IntlProvider locale="en-US" messages={locale}>
            <ProductList
              products={[
                {
                  id: 1,
                  name: 'test',
                  key: 'test',
                  price: 0,
                  currency: 'USD',
                  weight_unit: 'KG',
                  packaging: 'pack',
                  notes: 'pack',
                  image:
                    'https://burst.shopifycdn.com/photos/tea-cup-with-hot-peppers-and-yellow-tomatoes-on-red.jpg?width=240&format=pjpg&exif=1&iptc=1',
                },
              ]}
            />
          </IntlProvider>
        </RecoilRoot>
      </ThemeProvider>,
    );

    expect(temp).toMatchSnapshot();
    expect(temp.queryByTestId('productList')).toBeTruthy();
  });

  it('renders expected number of products', function () {
    const temp = render(
      <ThemeProvider>
        <RecoilRoot>
          <IntlProvider locale="en-US" messages={locale}>
            <ProductList
              products={[
                {
                  id: 1,
                  name: 'test',
                  key: 'test',
                  price: 0,
                  currency: 'USD',
                  weight_unit: 'KG',
                  packaging: 'pack',
                  notes: 'pack',
                  image:
                    'https://burst.shopifycdn.com/photos/tea-cup-with-hot-peppers-and-yellow-tomatoes-on-red.jpg?width=240&format=pjpg&exif=1&iptc=1',
                },
              ]}
            />
          </IntlProvider>
        </RecoilRoot>
      </ThemeProvider>,
    );

    const getTestId = temp.queryByTestId('productListData')!;
    expect(getTestId.props.data.length).toBe(1);
  });
});
