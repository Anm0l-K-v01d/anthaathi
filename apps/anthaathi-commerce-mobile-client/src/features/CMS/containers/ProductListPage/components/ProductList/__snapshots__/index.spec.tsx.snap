// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList should render the item 1`] = `
<View
  style={
    {
      "marginHorizontal": 5,
      "marginTop": 10,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "paddingBottom": 100,
      }
    }
    data={
      [
        [
          {
            "currency": "USD",
            "image": "https://burst.shopifycdn.com/photos/tea-cup-with-hot-peppers-and-yellow-tomatoes-on-red.jpg?width=240&format=pjpg&exif=1&iptc=1",
            "key": "test",
            "name": "test",
            "notes": "pack",
            "packaging": "pack",
            "price": 0,
            "weight_unit": "KG",
          },
        ],
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="productList"
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 15,
            }
          }
        >
          <View
            style={
              {
                "backgroundColor": "#f0f0f0",
                "borderColor": "#e7e7e7",
                "borderRadius": 12,
                "borderWidth": 1,
                "marginVertical": 5,
                "width": "48%",
              }
            }
          >
            <View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                  }
                }
              >
                <Image
                  source={
                    {
                      "uri": "https://burst.shopifycdn.com/photos/tea-cup-with-hot-peppers-and-yellow-tomatoes-on-red.jpg?width=240&format=pjpg&exif=1&iptc=1",
                    }
                  }
                  style={
                    {
                      "backgroundColor": "#fff",
                      "borderRadius": 4,
                      "height": 200,
                      "marginHorizontal": 10,
                      "width": "100%",
                    }
                  }
                />
                <View
                  style={
                    {
                      "alignItems": "flex-start",
                      "backgroundColor": "#f0f0f0",
                      "paddingHorizontal": 10,
                      "paddingVertical": 5,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "flex-start",
                        "flexDirection": "row",
                        "justifyContent": "space-between",
                        "width": 190,
                      }
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "#000000",
                            "fontFamily": "System",
                            "fontWeight": "400",
                          },
                          {
                            "textAlign": "left",
                          },
                          [
                            {
                              "textAlign": "left",
                            },
                            {
                              "color": "rgba(0, 0, 0, 0.87)",
                              "fontFamily": "System",
                              "fontWeight": "500",
                              "writingDirection": "ltr",
                            },
                            [
                              {
                                "fontSize": 20,
                                "letterSpacing": 0.15,
                                "lineHeight": 30,
                                "marginVertical": 2,
                              },
                              {
                                "fontSize": 12,
                                "fontWeight": "bold",
                                "marginRight": 5,
                              },
                            ],
                          ],
                        ]
                      }
                    >
                      test
                    </Text>
                    <Text
                      allowFontScaling={false}
                      selectable={false}
                      style={
                        [
                          {
                            "color": "#364A15",
                            "fontSize": 18,
                          },
                          {
                            "marginTop": 5,
                          },
                          {
                            "fontFamily": "Entypo",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          {},
                        ]
                      }
                    >
                      Ôáç
                    </Text>
                  </View>
                  <Text
                    style={
                      [
                        {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        {
                          "textAlign": "left",
                        },
                        {
                          "color": "#808080",
                          "fontSize": 12,
                          "fontWeight": "400",
                        },
                      ]
                    }
                  >
                    Dorne
                  </Text>
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "marginTop": 5,
                      }
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "#000000",
                            "fontFamily": "System",
                            "fontWeight": "400",
                          },
                          {
                            "textAlign": "left",
                          },
                          {
                            "color": "#008D3E",
                            "fontSize": 14,
                            "fontWeight": "400",
                          },
                        ]
                      }
                    >
                      $0.00
                    </Text>
                    <Text
                      style={
                        [
                          {
                            "color": "#000000",
                            "fontFamily": "System",
                            "fontWeight": "400",
                          },
                          {
                            "textAlign": "left",
                          },
                          {
                            "color": "#808080",
                            "fontSize": 12,
                            "fontWeight": "400",
                          },
                        ]
                      }
                    >
                       / pack
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
